pintc/src/asm_gen/asm_builder.rs

```diff
@@ -1090,11 +1090,22 @@ impl<'a> AsmBuilder<'a> {
             ));
         }
 
+        // check it's a prime op
+        let tuple_expr = if let Expr::UnaryOp {
+            op: UnaryOp::NextState,
+            expr,
+            ..
+        } = tuple.get(contract)
+        {
+            expr
+        } else {
+            tuple
+        };
+
```

Comment:
@otrho While I add e2e tests, I wanted to double-check the fix logic with you.


https://github.com/essential-contributions/pint/blob/8ea90ce77b392397e0d6c7b298e9c13500f44905/pintc/src/predicate/transform/lower.rs#L1074-L1102
Here the `UnaryOp::NextState` is stuffed into the `Expr::TupleFieldAccess` with `*old_accessed_key = op_key;`. I _believe_ that is correct, as we want to know that the field access is a prime op. If that's the case, then we want to make sure to dig into the field access and see if the `Expr::Tuple` is inside of the prime op before attempting to access the fields of the tuple.

Does that seem correct? Or should the prime op not be inside the `Expr::TupleFieldAccess` and be preserved elsewhere?

Reply:
Yeah, this is a bit unintuitive.  I think the prime op should be as close to the base storage value as possible, so it would go inside the field access.  This way, when we're fetching storage into state, it's known without any fuss whether we want pre- or post-.

So the type of the field access is the type of the field, and if the field is a prime op, it should be transparent.  I haven't looked deeper into why we're getting those internal errors though, I'll leave that to you. :slightly_smiling_face:

Reply:
> Yeah, this is a bit unintuitive. I think the prime op should be as close to the base storage value as possible, so it would go inside the field access. This way, when we're fetching storage into state, it's known without any fuss whether we want pre- or post-.

So the current setup is correct then 👍 

> So the type of the field access is the type of the field, and if the field is a prime op, it should be transparent. I haven't looked deeper into why we're getting those internal errors though, I'll leave that to you. 🙂

We were getting the error because the prime op was the field access, as intended. When in the `asm_builder`, we just didn't check if it was a prime op first. So instead of digging into the prime op to get the tuple that the prime op references, we panicked. The fix above should remedy that.

Here's where we got the error before I made the changes
```
let Type::Tuple { ref fields, .. } = tuple.get_ty(contract) else {
    return Err(
    handler.emit_internal_err("type must exist and be a tuple type", empty_span())
    );
```